chainio/dispatcher.go

```diff
@@ -0,0 +1,290 @@
+func (b *BlockbeatDispatcher) dispatchBlocks(
+	blockEpochs *chainntnfs.BlockEpochEvent) {
+
+	defer b.wg.Done()
+	defer blockEpochs.Cancel()
+
+	for {
+		select {
+		case blockEpoch, ok := <-blockEpochs.Epochs:
+			if !ok {
+				clog.Debugf("Block epoch channel closed")
+
+				return
+			}
+
+			clog.Infof("Received new block %v at height %d, "+
+				"notifying consumers...", blockEpoch.Hash,
+				blockEpoch.Height)
+
+			// Record the time it takes the consumer to process
+			// this block.
+			start := time.Now()
+
+			// Update the current block epoch.
+			b.beat = NewBeat(*blockEpoch)
+
+			// Notify all consumers.
+			err := b.notifyQueues()
+
+			// It's critical that the subsystems can process blocks
+			// correctly and timely, if an error returns, we'd
+			// gracefully shutdown lnd to bring attentions.
+			if err != nil {
+				b.log().Criticalf("Notify block failed: %v",
+					err)
+			}
+
+			b.log().Infof("Notified all consumers on new block "+
+				"in %v", time.Since(start))
+
+		case <-b.quit:
+			b.log().Debugf("BlockbeatDispatcher quit signal " +
+				"received")
+
+			return
+		}
+	}
+}
+
+// notifyQueues notifies each queue concurrently about the latest block epoch.
+func (b *BlockbeatDispatcher) notifyQueues() error {
+	eg := &errgroup.Group{}
+
+	// Notify each queue in goroutines.
+	for qid, consumers := range b.consumerQueues {
+		b.log().Debugf("Notifying queue=%d with %d consumers", qid,
+			len(consumers))
+
+		// Notify each queue concurrently.
+		eg.Go(func() error {
+			// Notify each consumer in this queue sequentially.
+			err := DispatchSequential(b.beat, consumers)
+
+			// Exit early if there's no error.
+			if err == nil {
+				return nil
+			}
+
+			return fmt.Errorf("queue=%d got err: %w", qid, err)
+		})
+	}
+
+	// Wait for all consumers in each queue to finish.
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+
+	b.log().Debugf("Notified all queues")
+
+	return nil
+}
+
+// DispatchSequential takes a list of consumers and notify them about the new
+// epoch sequentially. It requires the consumer to finish processing the block
+// within the specified time, otherwise a timeout error is returned.
+func DispatchSequential(b Blockbeat, consumers []Consumer) error {
+	for _, c := range consumers {
+		// Send the beat to the consumer.
+		err := notifyAndWait(b, c, DefaultProcessBlockTimeout)
+		if err != nil {
+			b.logger().Errorf("Failed to process block: %v", err)
+
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DispatchConcurrent notifies each consumer concurrently about the blockbeat.
+// It requires the consumer to finish processing the block within the specified
+// time, otherwise a timeout error is returned.
+func DispatchConcurrent(b Blockbeat, consumers []Consumer) error {
+	eg := &errgroup.Group{}
+
+	// Notify each queue in goroutines.
+	for _, c := range consumers {
+		// Notify each consumer concurrently.
+		eg.Go(func() error {
+			// Send the copy of the beat to the consumer.
```

Comment:
Nit: Outdated comment, because we do not deepCopy anymore ?

Reply:
yeah removed